### PR TITLE
fix telnet trace times is always 1

### PR DESCRIPTION
**Pull request by LiZhenNet**
## What is the purpose of the change

old:
```java

 if (service == null || service.length() == 0) {
            service = parts.length > 0 ? parts[0] : null;
            method = parts.length > 1 ? parts[1] : null;
        } else {
            method = parts.length > 0 ? parts[0] : null;
        }
        if (StringUtils.isInteger(method)) {
            times = method;
            method = null;
        } else {
            times = parts.length > 2 ? parts[2] : "1";
        }
```

Usually we use telnet  in order，first，we invoke ”cd xxxService“ to change the path，then we invoke "trace  xxxMethod 10" to trace a method 。

If we invoke like this "trace  xxxMethod 10" and the service is not equals null（we have changed the path） , times will not change the value  ，It‘s always 1。

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by LiZhenNet** [comment]
@beiwei30  Can you help me review it？

**Comment by khanimteyaz** [comment]
@beiwei30 I would like to review this. would you help me to understand by functionality wise what @LiZhenNet to trying to achieve here?

**Comment by LiZhenNet** [comment]
> @beiwei30 I would like to review this. would you help me to understand by functionality wise what @LiZhenNet to trying to achieve here?

I try to fix a bug about trace command in telnet。You can see “What is the purpose of the change”

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/TraceTelnetHandler.java [hunk 1]
```diff
@@ -45,17 +45,18 @@ public String telnet(Channel channel, String message) {
         String[] parts = message.split("\\s+");
         String method;
         String times;
+        // message like : XxxService , XxxService 10 , XxxService xxxMethod , XxxService xxxMethod 10
         if (service == null || service.length() == 0) {
```
Comment (khanimteyaz) [inline dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/TraceTelnetHandler.java]: I think 
`if (StringUtils.isEmpty(service))`
be better.

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/TraceTelnetHandler.java [hunk 2]
```diff
@@ -45,17 +45,18 @@ public String telnet(Channel channel, String message) {
...
             service = parts.length > 0 ? parts[0] : null;
             method = parts.length > 1 ? parts[1] : null;
-        } else {
+            times = parts.length > 2 ? parts[2] : "1";
```
Comment (khanimteyaz) [inline dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/TraceTelnetHandler.java]: is there any chance parts[2] can be having non numeric number?

**Comment by LiZhenNet** [comment]
I have add a commit c833771 , to fix https://github.com/apache/incubator-dubbo/issues/3105

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/InvokeTelnetHandler.java [hunk 3]
```diff
@@ -99,13 +97,10 @@ private static boolean isMatch(Method method,List<Object> args, Class<?>[] param
...
                     return false;
                 }
-            } else if (arg instanceof Map) {
-                String name = (String) ((Map<?, ?>) arg).get("class");
-                Class<?> cls = arg.getClass();
-                if (name != null && name.length() > 0) {
-                    cls = ReflectUtils.forName(name);
-                }
-                if (!type.isAssignableFrom(cls)) {
+            } else if (arg instanceof JSONObject) {
```
Comment (gudegg) [inline dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/InvokeTelnetHandler.java]:
```java
@Data
public class User1 implements Serializable {
    private static final long serialVersionUID = -4729079781904156102L;
    private String name;
}
```
```java
@Data
public class User2 implements Serializable {
    private static final long serialVersionUID = -2112637973209566052L;
    private String name;
}
```
```java
public interface XXXService{
String hello(User1 user);
String hello(User2 user);

}
```
May find not the correct overload method; I think  must **class** param when the method is overloaded.

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/InvokeTelnetHandler.java [hunk 4]
```diff
@@ -99,13 +97,10 @@ private static boolean isMatch(Method method,List<Object> args, Class<?>[] param
...
+                try {
+                    ((JSONObject) arg).toJavaObject(type);
```
Comment (gudegg) [inline dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/InvokeTelnetHandler.java]:
```java
@Data
public class User1 implements Serializable {
    private static final long serialVersionUID = -4729079781904156102L;
    private String name;
}
```
```java
@Data
public class User2 implements Serializable {
    private static final long serialVersionUID = -2112637973209566052L;
    private String name;
}
```
```java
public interface XXXService{
String hello(User1 user);
String hello(User2 user);

}
```
May find not the correct overload method; I think  must **class** param when the method is overloaded.
telnet: invoke xxxService.hello({"name":"xxxx"})

Comment (LiZhenNet) [reply dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/InvokeTelnetHandler.java]: I have fix it

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/InvokeTelnetHandler.java [hunk 5]
```diff
@@ -101,12 +100,19 @@ private static boolean isMatch(Method method,List<Object> args, Class<?>[] param
                 }
             } else if (arg instanceof Map) {
                 String name = (String) ((Map<?, ?>) arg).get("class");
-                Class<?> cls = arg.getClass();
-                if (name != null && name.length() > 0) {
-                    cls = ReflectUtils.forName(name);
-                }
-                if (!type.isAssignableFrom(cls)) {
-                    return false;
+                if (StringUtils.isNotEmpty(name)) {
```
Comment (beiwei30) [inline dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/InvokeTelnetHandler.java]: 这里校验是不是太严格了，我觉得简单一点，没有 `class` 的时候返回 `true` 就 好了。

**Pull request merged**